tests/eos-test-utils.c

```diff
@@ -0,0 +1,1960 @@
+
+  return g_file_query_exists (config, NULL);
+}
+
+gboolean
+eos_test_subserver_update (EosTestSubserver *subserver,
+                           GError **error)
+{
+  GFile *repo = subserver->repo;
+  g_autoptr(GDateTime) timestamp = NULL;
+  g_auto(CmdStuff) cmd = CMD_STUFF_CLEARED;
+
+  if (!create_directory (repo, error))
+    return FALSE;
+
+  if (!repo_config_exists (repo))
+    {
+      if (!ostree_init (repo,
+                        REPO_ARCHIVE_Z2,
+                        &cmd,
+                        error))
+        return FALSE;
+      if (!cmd_stuff_ensure_ok (&cmd, error))
+        return FALSE;
+    }
+
+  if (!update_commits (subserver, error))
+    return FALSE;
+
+  if (subserver->branch_file_timestamp != NULL)
+    timestamp = g_date_time_ref (subserver->branch_file_timestamp);
+  else
+    timestamp = g_date_time_new_now_utc ();
+  if (!generate_branch_file (repo,
+                             subserver->devices,
+                             subserver->keyid,
+                             timestamp,
+                             error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+eos_test_server_dispose_impl (EosTestServer *server)
+{
+  g_clear_object (&server->root);
+  g_clear_pointer (&server->subservers, g_ptr_array_unref);
+}
+
+static void
+eos_test_server_finalize_impl (EosTestServer *server)
+{
+  g_free (server->url);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_TEST_SERVER,
+                       EosTestServer,
+                       eos_test_server,
+                       eos_test_server_dispose_impl,
+                       eos_test_server_finalize_impl)
+
+static gboolean
+run_httpd (GFile *served_root,
+	   gchar **out_url,
+	   GError **error)
+{
+  g_auto(CmdStuff) cmd = CMD_STUFF_CLEARED;
+  guint16 port;
+
+  if (!ostree_httpd (served_root,
+                     &port,
+                     &cmd,
+                     error))
+    return FALSE;
+  if (!cmd_stuff_ensure_ok (&cmd, error))
+    return FALSE;
+
+  *out_url = g_strdup_printf ("http://127.0.0.1:%d", port);
+  return TRUE;
+}
+
+static GFile *
+get_main_tree_root (GFile *main_root)
+{
+  return g_file_get_child (main_root, "trees");
+}
+
+static GFile *
+get_main_served_root (GFile *main_root)
+{
+  return g_file_get_child (main_root, "served");
+}
+
+static gboolean
+setup_subservers (GPtrArray *subservers,
+		  GFile *main_root,
+		  GError **error)
+{
+  g_autoptr(GFile) tree_root = NULL;
+  g_autoptr(GFile) served_root = NULL;
+  guint idx;
+
+  tree_root = get_main_tree_root (main_root);
+  served_root = get_main_served_root (main_root);
+  for (idx = 0; idx < subservers->len; ++idx)
+    {
+      EosTestSubserver *subserver = EOS_TEST_SUBSERVER (g_ptr_array_index (subservers, idx));
+      g_autoptr(GFile) subtree = g_file_get_child (tree_root, subserver->ostree_path);
+      g_autoptr(GFile) subserved = g_file_get_child (served_root, subserver->ostree_path);
+
+      g_set_object (&subserver->repo, subserved);
```

Comment:
Should we do anything here if the files obtained from `g_file_get_child` don't exist?